---
title: "Integrating datasets"
output: md_document
date: "2023-10-25"
---

# Intro to data integration

A recurrent theme in -omics work is the notion of "batch effects". This term originates from the days of microarrays, when there would be slight differences in annealing times and/or affinities between runs of microarrays. "Batch effects" has now become a blanket colloquial term to generally mean anything related to a technical artifact that arises between samples or runs. The key is that it is not a true biological signal, but instead something that should be "normalized away" to allow real biological signals to shine.

Statistically, it is straightforward to control for a batch effect: if you are using a linear model, for example, you could include the "batch" as a covariate. The challenge really arises with high-dimensional approaches such as principal component analysis or visualizing data in a UMAP - often times, when batch effects are present, visualizations are distorted or largely driven by these batch effects. There are well-documented strategies for removing batch effect in microarrays or bulk RNAseq data, which we will not discuss here. 

For single-cell RNAseq analysis, we can use one simple approach called "data integration" that is robustly implemented in Seurat. To demonstrate this, we will use healthy donor PBMC that have been generated with either 10X Genomics 3' v2 chemistry or 5' v1 chemistry. The chemistry, in this case, is the batch effect we are trying to remove. 

# Prerequisites 

We will need to download some dataset from the Cillo Laboratory [tutorial_datasets](https://github.com/CilloLaboratory/tutorial_datasets) github page.

If you navigate to the link, click the code button on the top right, and click "Download ZIP" it will download the whole repository including directories containing the 10X scRNAseq datasets.

For this tutorial, we will be using the *PBMC_integration_datasets* directory and enclosed files. 

# Load packages 

```{r load_packages}

library(Seurat)
library(dplyr)
library(patchwork)

```

```{r read_data}

full_path <- "/home/rstudio/docker_rstudio/data/tutorial_datasets-main/PBMC_integration_datasets/"

raw_files <- list.files(paste(full_path))

# We don't want to read in the metadata as a 10X dataset
# Let's remove it from the raw_files
raw_files <- raw_files[!grepl("txt",raw_files)]
raw_files

# Create list for Seurat object
raw_list <- vector("list",length=length(raw_files))
raw_list # empty list of length 4

# Loop to read filtered feature barcode matrices into R data into R
for (i in 1:length(raw_list)) {
  
  raw_list[[i]] <- Read10X(paste(full_path,raw_files[i],sep=""))
  
}

# Name each item in the list by its title
names(raw_list) <- raw_files
names(raw_list)

```

# Read in metadata and create Seurat objects

```{r create_ser}

# Read in the metadata
metadata_file <- readr::read_tsv(paste(full_path,"pbmc_integration_metadata.txt",
                                       sep=""))

# Make sure the order of names matches the raw data list 
identical(metadata_file$sample_id,names(raw_list))

# Create an empty list of Seurat objects
ser_list <- vector("list",length=length(raw_list))

# Loop to Create individual Seurat objects and include metadata 
for (i in 1:length(ser_list)) {
  
  # Create Seurat Object
  ser_list[[i]] <- CreateSeuratObject(raw_list[[i]])
  
  # Create metadata for each object
  meta_add <- data.frame(matrix(data=NA,
                                nrow=ncol(ser_list[[i]]),
                                ncol=ncol(metadata_file))
                         )
  colnames(meta_add) <- colnames(metadata_file)
  rownames(meta_add) <- colnames(ser_list[[i]])
  meta_add[,1] <- metadata_file[i,1] %>% pull()
  meta_add[,2] <- metadata_file[i,2] %>% pull()
  meta_add[,3] <- metadata_file[i,3] %>% pull()
  
  # Add Metadata
  ser_list[[i]] <- AddMetaData(ser_list[[i]],metadata=meta_add)

}

# Check out our list of Seurat objects
ser_list

```

# Merge list of Seurat objects into a single seurat object 

```{r merge_ser}

ser_merged <- merge(ser_list[[1]],ser_list[2:length(ser_list)])

ser_merged@meta.data$sample_id %>%
  table()

ser_merged@meta.data %>%
  select(sample_id,chemistry) %>%
  table()

```

# QC filtering

Now that we have out combined object, let's perform some filtering before we integrate.

```{r qc_filtering}

ser_merged[["percent_mt"]] <- PercentageFeatureSet(ser_merged, pattern = "^MT-")

# Check out metadata now
head(ser_merged@meta.data)

# We can also plot values from metadata
VlnPlot(ser_merged, features = c("nFeature_RNA", "nCount_RNA", "percent_mt"), ncol = 3)

# Filtering is often dataset specific
# Here, we want more than 200 and less than 3000 features
# Plus % mito < 20%
ser_merged <- subset(ser_merged, subset = nFeature_RNA > 200 & nFeature_RNA < 3000 & percent_mt < 20)

# Check out the cells left 
ser_merged@meta.data %>%
  select(sample_id,chemistry) %>%
  table()

```

# Standard workflow

We will run the standard workflow first to demonstrate the chemistry specific batch effects. 

```{r standard}

# Store the filtered and un-manipulated dataset
ser_merged_save <- ser_merged

ser_merged <- ser_merged %>% 
    NormalizeData(.) %>%
    FindVariableFeatures(.) %>%
    ScaleData(.) %>%
    RunPCA(.) 

ElbowPlot(ser_merged)

ser_merged <- ser_merged %>%
    RunUMAP(.,dims=1:10) %>%
    FindNeighbors(.,dims=1:10) %>%
    FindClusters(.,res=c(0.3,0.5,0.7))

Idents(ser_merged) <- "RNA_snn_res.0.3"

DimPlot(ser_merged,group.by="chemistry",label=T)

DimPlot(ser_merged,group.by="RNA_snn_res.0.3",split.by="chemistry",label=T)

# Check out chemistry-specific clusters
ser_merged@meta.data %>%
  select(RNA_snn_res.0.3,chemistry) %>%
  table()

```

# Batch effects

Clearly, there are strong batch effects associated with the chemistry used to generated the data. Let's look at how we can correct these batch effects by integrating the datasets. 

We will use the SCTransform and rpca approaches described by the Satija Lab in Seurat. More info on integration approaches is available [here](https://satijalab.org/seurat/articles/integration_rpca#modifying-the-strength-of-integration-1).

```{r integration}

ser_merged_split <- SplitObject(ser_merged, split.by = "chemistry")

# lapply applies the FUN (function) to each member of the list and returns a list
ser_merged_split <- lapply(X = ser_merged_split, FUN = SCTransform, method = "glmGamPoi")

# Identify the conserved features for integration 
features <- SelectIntegrationFeatures(object.list = ser_merged_split, 
                                      nfeatures = 2000
                                      )

# Prepare for integration 
ser_merged_split <- PrepSCTIntegration(object.list = ser_merged_split, anchor.features = features)
ser_merged_split <- lapply(X = ser_merged_split, FUN = RunPCA, features = features)

# Find anchors and integrate
anchors_to_use <- FindIntegrationAnchors(object.list = ser_merged_split,
                                         normalization.method = "SCT",
                                         anchor.features = features,
                                         dims = 1:30, 
                                         reduction = "rpca",
                                         k.anchor = 20)

ser_integrated <- IntegrateData(anchorset = anchors_to_use, 
                                normalization.method = "SCT", 
                                dims = 1:30)

```

# Run PCA, UMAP and clustering on integrated data 

```{r integrated}

ser_integrated <- RunPCA(ser_integrated,verbose = FALSE)

ElbowPlot(ser_integrated)

ser_integrated <- RunUMAP(ser_integrated, reduction = "pca", dims = 1:10) %>%
  FindNeighbors(.,dims=1:10) %>%
  FindClusters(.,res=c(0.3,0.5,0.7))

```

# Visualize integrated data 

```{r int_viz}

colnames(ser_integrated@meta.data)
# Notice the clustering results now have the preface "integrated"
# Instead of RNA 

# A new assay is present called "integrated"
ser_integrated
ser_integrated@assays

# View by chemistry and clusters
DimPlot(ser_integrated,group.by="chemistry")

DimPlot(ser_integrated,group.by="integrated_snn_res.0.3")

# View chemistry across clusters - much better 
ser_integrated@meta.data %>%
  select(integrated_snn_res.0.3,chemistry) %>%
  table()

```

# Identifying DEGs

Now that we have integrated the datasets, we can look for differentially expressed genes across the clusters.

It's important to now switch back to the RNA assay. We do not want to use only the 2000 highly variable genes contained in the integrated assay to find differentially expressed genes. 

```{r degs}

DefaultAssay(ser_integrated) <- "RNA"
Idents(ser_integrated) <- "integrated_snn_res.0.3"

# Using logfc.thresold at 0.5 to speed this up 
all_degs <- FindAllMarkers(ser_integrated,logfc.threshold = 0.5)

all_degs %>%
  group_by(cluster) %>%
  top_n(5,avg_log2FC) %>%
  arrange(desc(avg_log2FC))

```

# Conclusions 

We have now walked through a method to integrate two datasets that have a known batch effect like chemistry. In practice, a batch effect can be anything from a patient specific sample to differences in how data was generated or differences in sequencing depth between samples.

# Saving data 

```{r saving}

# saveRDS(ser_integrated,file=paste(full_path,"PBMC_integration_final_output_231024.rds",sep=""))

```
